android/src/main/java/com/mapbox/reactnativemapboxgl/ReactNativeMapboxGLModule.java

```diff
@@ -59,87 +119,414 @@ public String getName() {
+        pixelRatio = pixelRatio < 1.5f ? 1.0f : 2.0f;
+
+        ReadableArray boundsArray = options.getArray("bounds");
+        LatLngBounds bounds = new LatLngBounds.Builder()
+                .include(new LatLng(boundsArray.getDouble(0), boundsArray.getDouble(1)))
+                .include(new LatLng(boundsArray.getDouble(2), boundsArray.getDouble(3)))
+                .build();
+
+        final OfflineTilePyramidRegionDefinition regionDef = new OfflineTilePyramidRegionDefinition(
+                options.getString("styleURL"),
+                bounds,
+                options.getDouble("minZoomLevel"),
+                options.getDouble("maxZoomLevel"),
+                pixelRatio
+        );
+
+        byte [] metadata;
+
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            ObjectOutputStream oos = new ObjectOutputStream(bos);
+            oos.writeObject(options.getString("name"));
+            oos.writeObject(options.getString("metadata"));
+            oos.close();
+            metadata = bos.toByteArray();
+        } catch (IOException e) {
+            promise.reject(e);
+            return;
+        }
+
+        final ReactNativeMapboxGLModule _this = this;
+        final byte [] _metadata = metadata;
+        mainHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                OfflineManager.getInstance(context.getApplicationContext()).createOfflineRegion(
+                        regionDef,
+                        _metadata,
+                        new OfflineManager.CreateOfflineRegionCallback() {
+                            @Override
+                            public void onCreate(OfflineRegion offlineRegion) {
+                                OfflineRegionProgressObserver observer = new OfflineRegionProgressObserver(_this, offlineRegion, null);
+                                offlinePackObservers.add(observer);
+                                offlineRegion.setObserver(observer);
+                                offlineRegion.setDownloadState(OfflineRegion.STATE_ACTIVE);
+                                promise.resolve(null);
+                            }
+
+                            @Override
+                            public void onError(String error) {
+                                promise.reject(new JSApplicationIllegalArgumentException(error));
+                            }
+                        }
+                );
+            }
+        });
+    }
+
+    // Offline pack removal
+
+    class OfflineRegionDummyObserver implements OfflineRegion.OfflineRegionObserver {
+
+        @Override
+        public void onStatusChanged(OfflineRegionStatus status) {
+
+        }
+
+        @Override
+        public void onError(OfflineRegionError error) {
+
+        }
+
+        @Override
+        public void mapboxTileCountLimitExceeded(long limit) {
+
+        }
```

Comment:
After removing a region, mapbox still calls methods on the observer, so I just swapped the OfflineRegionProgressObserver with a dummy one. Later I found that this doesn't work as the observer change doesn't go into effect, so I added an `invalidated` flag in OfflineRegionProgressObserver. Therefore, I guess this class can go.